File: parser/props/DNASignGraphic.py
import struct

import DNANode


class DNASignGraphic(DNANode.DNANode):
    PROP_CODE = 8

    def __init__(self, name):
        DNANode.DNANode.__init__(self, name)

        self.code = ''
        self.color = (1, 1, 1, 1)
        self.width = 0
        self.height = 0
        self.bDefaultColor = True

    def getCode(self):
        return self.code

    def setCode(self, code):
        self.code = code

    def setColor(self, color):
        self.color = color
        self.bDefaultColor = False

    def getColor(self):
        return self.color

    def setWidth(self, width):
        self.width = width

    def getWidth(self):
        return self.width

    def setHeight(self, height):
        self.height = height

    def getHeight(self):
        return self.height

    def debug(self, message):
        if self.verbose:
            print 'DNASignGraphic:', message

    def traverse(self, recursive=True, verbose=False):
        data = DNANode.DNANode.traverse(self, recursive=False, verbose=verbose)

        code = self.getCode()
        data += struct.pack('>B', len(code))  # Code length
        self.debug('packing... code length: {0}'.format(len(code)))
        data += code  # Code
        self.debug('packing... code: {0}'.format(code))

        for component in self.getColor():
            data += struct.pack('>f', component)  # Color
            self.debug('packing... color: {0}'.format(component))

        data += struct.pack('f', self.getWidth())  # Width
        self.debug('packing... width: {0}'.format(self.getWidth()))

        data += struct.pack('f', self.getHeight())  # Height
        self.debug('packing... height: {0}'.format(self.getHeight()))

        data += struct.pack('?', self.bDefaultColor)  # bDefaultColor?
        self.debug('packing... bDefaultColor?: {0}'.format(self.bDefaultColor))

        if recursive:
            for child in self.children:
                data += child.traverse(recursive=True, verbose=verbose)

        return data

File: reader/DNASignGraphic.cpp
#include "DNASignGraphic.h"
#include <decalEffect.h>
#include "DNASign.h"

DNASignGraphic::DNASignGraphic(string name) : DNANode(name), m_color(new LVector4f(1)) {}
DNASignGraphic::~DNASignGraphic(void) {}

void DNASignGraphic::make_from_dgi(DatagramIterator& dgi, DNAStorage* store)
{
	DNANode::make_from_dgi(dgi, store);
	m_code = dgi_extract_string8(dgi);
	m_color = dgi_extract_color(dgi);
	m_width = dgi.get_int16() / 100;
	m_height = dgi.get_int16() / 100;
	m_b_default_color = dgi.get_bool();
};

void DNASignGraphic::traverse(NodePath& np, DNAStorage* store)
{
	NodePath* result = store->find_node(m_code);
	if (result->is_empty())
	{
		stringstream ss;
		ss << "code " << m_code << " not in storage!";
		cout << "DNASignGraphic(fatal): " << ss.str() << endl;
		throw ss.str();
	}

	NodePath _np = result->copy_to(np);
	_np.set_scale(*m_scale);

	DNASign* p = reinterpret_cast<DNASign*>(m_parent);
	LVector3f* s = p->get_scale();
	_np.set_scale(_np, *s);

	_np.set_pos(*m_pos);
	_np.set_hpr(*m_hpr);
	
	for (comp_vec_t::iterator child = m_children.begin(); child != m_children.end(); child++)
	{
		(*child)->traverse(_np, store);
	};
};

void DNASignGraphic::set_code(const string& code)
{
	m_code = code;
};

string DNASignGraphic::get_code()
{
	return m_code;
};

void DNASignGraphic::set_color(LVector4f* color)
{
	m_color = color;
};

LVector4f* DNASignGraphic::get_color()
{
	return m_color;
};

void DNASignGraphic::set_width(float w)
{
	m_width = w;
};

float DNASignGraphic::get_width()
{
	return m_width;
};

void DNASignGraphic::set_height(float h)
{
	m_height = h;
};

float DNASignGraphic::get_height()
{
	return m_height;
};
